velox/functions/sparksql/String.h

```diff
@@ -583,4 +583,122 @@ struct OverlayVarbinaryFunction : public OverlayFunctionBase {
+///
+///   Returns a new translated string. It translates the character in ``string``
+///   by a character in ``replace``. The character in ``replace`` is
+///   corresponding to the character in ``match``. The translation will
+///   happen when any character in ``string`` matching with the character in the
+///   ``match``.
+template <typename T>
+struct TranslateFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // ASCII input always produces ASCII result.
+  static constexpr bool is_default_ascii_behavior = true;
+
+  FOLLY_ALWAYS_INLINE static std::function<int(const char*)> utf8CharCounter() {
+    return [](const char* input) { return utf8proc_char_length(input); };
+  }
+  FOLLY_ALWAYS_INLINE static std::function<int(const char*)>
+  asciiCharCounter() {
+    return [](const char*) { return 1; };
+  }
+
+  using MapType = folly::F14FastMap<StringView, StringView>;
+  MapType dictionary_;
+
+  bool isConstantTranslateDictionary = false;
+
+  FOLLY_ALWAYS_INLINE MapType buildDictionary(
+      const arg_type<Varchar>& match,
+      const arg_type<Varchar>& replace,
+      std::function<int(const char* input)> charCounter) {
+    MapType dictionary;
+    int i = 0;
+    int j = 0;
+    while (i < match.size()) {
+      StringView replacePiece;
+      // If match's character size is larger than replace's, the execessive
+      // characters in match will be removed from input string.
+      if (j >= replace.size()) {
+        replacePiece = StringView("");
+      } else {
+        int replaceCharCount = charCounter(replace.data() + j);
+        replacePiece = StringView(replace.data() + j, replaceCharCount);
+        j = j + replaceCharCount;
+      }
+      int matchCharCount = charCounter(match.data() + i);
+      StringView matchPiece = StringView(match.data() + i, matchCharCount);
+      // Only consider the first occurrence of a character in match.
+      if (dictionary.find(matchPiece) == dictionary.end()) {
+        dictionary.emplace(matchPiece, replacePiece);
+      }
+      i = i + matchCharCount;
+    }
+    return dictionary;
+  }
+
+  FOLLY_ALWAYS_INLINE void initialize(
+      const core::QueryConfig& /*config*/,
+      const arg_type<Varchar>* /*string*/,
+      const arg_type<Varchar>* match,
+      const arg_type<Varchar>* replace) {
+    if (match != nullptr && replace != nullptr) {
+      dictionary_ = buildDictionary(*match, *replace, utf8CharCounter());
+      isConstantTranslateDictionary = true;
+    }
+  }
+
+  FOLLY_ALWAYS_INLINE void doTranslate(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& input,
+      std::function<int(const char* input)> charCounter) {
+    int k = 0;
+    while (k < input.size()) {
+      int inputCharCount = charCounter(input.data() + k);
+      StringView inputPiece = StringView(input.data() + k, inputCharCount);
+      auto it = dictionary_.find(inputPiece);
+      if (it == dictionary_.end()) {
+        result.append(inputPiece);
```

Comment:
Wouldn't appending one character at a time be very slow?

Reply:
Just aware of this issue. Fixed in the latest commit by reserving capacity.